File: _code-samples/send-a-trust-line-token/js/sendTrustLineToken.js
// IMPORTANT: This example sends a trust line token on the XRP Ledger Testnet
// from a preconfigured issuer to a holder. It creates a trust line from the
// holder to the issuer, sends a Payment of the issued currency, and verifies
// the resulting balances. The token sent here is "FOO".

import fs from 'fs'
import { execSync } from 'child_process'
import xrpl from 'xrpl'

// Set up client ----------------------
const client = new xrpl.Client('wss://s.altnet.rippletest.net:51233')
await client.connect()

// This step checks for the necessary setup data to run the tutorial.
// If missing, setup.js will fund the issuer and holder wallets and
// enable Default Ripple on the issuer.
if (!fs.existsSync('setup.json')) {
  console.log(`\n=== Tutorial setup data doesn't exist. Running setup script... ===\n`)
  execSync('node setup.js', { stdio: 'inherit' })
}

// Load preconfigured issuer and holder accounts.
const setupData = JSON.parse(fs.readFileSync('setup.json', 'utf8'))
const issuer = xrpl.Wallet.fromSeed(setupData.issuer.seed)
const holder = xrpl.Wallet.fromSeed(setupData.holder.seed)

console.log(`Issuer address: ${issuer.address}`)
console.log(`Holder address: ${holder.address}`)

const currencyCode = 'FOO'

// Create trust line ----------------------
// The holder authorizes the issuer to send up to the limit_amount of the
// issued currency to the holder's account.
console.log('\n=== Preparing TrustSet transaction ===\n')
const trustSetTx = {
  TransactionType: 'TrustSet',
  Account: holder.address,
  LimitAmount: {
    currency: currencyCode,
    issuer: issuer.address,
    value: '1000000000'
  }
}

xrpl.validate(trustSetTx)
console.log(JSON.stringify(trustSetTx, null, 2))

console.log('\n=== Submitting TrustSet transaction ===\n')
const trustSetResponse = await client.submitAndWait(trustSetTx, {
  wallet: holder,
  autofill: true
})
if (trustSetResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = trustSetResponse.result.meta.TransactionResult
  console.error('Error: Unable to create trust line:', resultCode)
  await client.disconnect()
  process.exit(1)
}
console.log('Trust line created from holder to issuer.')

// Send issued token ----------------------
// The issuer sends a Payment of the issued currency to the holder. The
// Amount object specifies the currency, issuer, and value.
const issueQuantity = '500'
console.log('\n=== Preparing Payment transaction ===\n')
const paymentTx = {
  TransactionType: 'Payment',
  Account: issuer.address,
  Destination: holder.address,
  Amount: {
    currency: currencyCode,
    issuer: issuer.address,
    value: issueQuantity
  }
}

xrpl.validate(paymentTx)
console.log(JSON.stringify(paymentTx, null, 2))

console.log('\n=== Submitting Payment transaction ===\n')
const paymentResponse = await client.submitAndWait(paymentTx, {
  wallet: issuer,
  autofill: true
})
if (paymentResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = paymentResponse.result.meta.TransactionResult
  console.error('Error: Unable to send issued token:', resultCode)
  await client.disconnect()
  process.exit(1)
}
console.log(`Sent ${issueQuantity} ${currencyCode} from issuer to holder.`)

// Verify balances ----------------------
// Confirm the holder received the token by reading its trust line balances,
// and confirm the issuer's outstanding obligation through gateway_balances.
console.log('\n=== Verifying balances ===\n')
const holderLines = await client.request({
  command: 'account_lines',
  account: holder.address,
  ledger_index: 'validated'
})
console.log('Holder trust lines:')
console.log(JSON.stringify(holderLines.result.lines, null, 2))

const issuerObligations = await client.request({
  command: 'gateway_balances',
  account: issuer.address,
  ledger_index: 'validated'
})
console.log('\nIssuer obligations:')
console.log(JSON.stringify(issuerObligations.result.obligations, null, 2))

await client.disconnect()

File: _code-samples/send-a-trust-line-token/py/send_trust_line_token.py
# IMPORTANT: This example sends a trust line token on the XRP Ledger Testnet
# from a preconfigured issuer to a holder. It creates a trust line from the
# holder to the issuer, sends a Payment of the issued currency, and verifies
# the resulting balances. The token sent here is "FOO".

import json
import os
import subprocess
import sys

from xrpl.clients import JsonRpcClient
from xrpl.models import AccountLines, GatewayBalances
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.models.transactions import Payment, TrustSet
from xrpl.transaction import submit_and_wait
from xrpl.wallet import Wallet

# Set up client ----------------------
client = JsonRpcClient("https://s.altnet.rippletest.net:51234")

# This step checks for the necessary setup data to run the tutorial.
# If missing, setup.py will fund the issuer and holder wallets and
# enable Default Ripple on the issuer.
if not os.path.exists("setup.json"):
    print("\n=== Tutorial setup data doesn't exist. Running setup script... ===\n")
    subprocess.run([sys.executable, "setup.py"], check=True)

# Load preconfigured issuer and holder accounts.
with open("setup.json") as f:
    setup_data = json.load(f)

issuer = Wallet.from_seed(setup_data["issuer"]["seed"])
holder = Wallet.from_seed(setup_data["holder"]["seed"])

print(f"Issuer address: {issuer.address}")
print(f"Holder address: {holder.address}")

currency_code = "FOO"

# Create trust line ----------------------
# The holder authorizes the issuer to send up to the limit_amount of the
# issued currency to the holder's account.
print("\n=== Preparing TrustSet transaction ===\n")
trust_set_tx = TrustSet(
    account=holder.address,
    limit_amount=IssuedCurrencyAmount(
        currency=currency_code,
        issuer=issuer.address,
        value="1000000000",
    ),
)

print(json.dumps(trust_set_tx.to_xrpl(), indent=2))

print("\n=== Submitting TrustSet transaction ===\n")
trust_set_response = submit_and_wait(trust_set_tx, client, holder)
result_code = trust_set_response.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"Error: Unable to create trust line: {result_code}")
    sys.exit(1)
print("Trust line created from holder to issuer.")

# Send issued token ----------------------
# The issuer sends a Payment of the issued currency to the holder. The
# amount object specifies the currency, issuer, and value.
issue_quantity = "500"
print("\n=== Preparing Payment transaction ===\n")
payment_tx = Payment(
    account=issuer.address,
    destination=holder.address,
    amount=IssuedCurrencyAmount(
        currency=currency_code,
        issuer=issuer.address,
        value=issue_quantity,
    ),
)

print(json.dumps(payment_tx.to_xrpl(), indent=2))

print("\n=== Submitting Payment transaction ===\n")
payment_response = submit_and_wait(payment_tx, client, issuer)
result_code = payment_response.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"Error: Unable to send issued token: {result_code}")
    sys.exit(1)
print(f"Sent {issue_quantity} {currency_code} from issuer to holder.")

# Verify balances ----------------------
# Confirm the holder received the token by reading its trust line balances,
# and confirm the issuer's outstanding obligation through gateway_balances.
print("\n=== Verifying balances ===\n")
holder_lines = client.request(AccountLines(
    account=holder.address,
    ledger_index="validated",
))
print("Holder trust lines:")
print(json.dumps(holder_lines.result["lines"], indent=2))

issuer_obligations = client.request(GatewayBalances(
    account=issuer.address,
    ledger_index="validated",
))
print("\nIssuer obligations:")
print(json.dumps(issuer_obligations.result.get("obligations", {}), indent=2))
